external/js/cothority/src/network/connection.ts

```diff
@@ -191,67 +188,66 @@ export class WebSocketConnection implements IConnection {
+    sendStream<T extends Message>(message: Message, reply: typeof Message):
+        Observable<[T, WebSocketAdapter]> {
+        return new Observable((sub) => {
 
-        if (!message.$type) {
-            onError(new Error(`message "${message.constructor.name}" is not registered`));
-            return;
-        }
+            if (!message.$type) {
+                sub.error(new Error(`message "${message.constructor.name}" is not registered`));
+                return;
+            }
 
-        if (!reply.$type) {
-            onError(new Error(`message "${reply}" is not registered`));
-            return;
-        }
+            const path = this.getURL() + "/" + this.service + "/" + message.$type.name.replace(/.*\./, "");
+            Log.lvl4(`Socket: new WebSocket(${path})`);
+            const ws = factory(path);
+            const bytes = Buffer.from(message.$type.encode(message).finish());
+            const timer = setTimeout(() => ws.close(1000, "timeout"), this.timeout);
+            ws.onOpen(() => {
+                Log.lvl3("Sending message to", path);
+                ws.send(bytes);
+            });
 
-        const path = `${this.getURL()}/${this.service}/${message.$type.name.replace(/.*\./, "")}`;
-        Log.lvl4(`Socket: new WebSocket(${path})`);
-        const ws = factory(path);
-        const bytes = Buffer.from(message.$type.encode(message).finish());
-        const timer = setTimeout(() => ws.close(1000, "timeout"), this.timeout);
-        ws.onOpen(() => {
-            Log.lvl3("Sending message to", path);
-            ws.send(bytes);
-        });
+            ws.onMessage((data: Buffer) => {
+                clearTimeout(timer);
+                const buf = Buffer.from(data);
+                Log.lvl4("Getting message with length:", buf.length);
+
+                try {
+                    const ret = reply.decode(buf) as T;
+                    sub.next([ret, ws]);
+                } catch (err) {
+                    if (err instanceof util.ProtocolError) {
+                        sub.error(err);
+                    } else {
+                        sub.error(
+                            new Error(`Other error: ${err}`),
+                        );
+                    }
+                }
+            });
 
-        ws.onMessage((data: Buffer) => {
-            clearTimeout(timer);
-            const buf = Buffer.from(data);
-            Log.lvl4("Getting message with length:", buf.length);
-
-            try {
-                const ret = reply.decode(buf) as T;
-                onMessage(ret, ws);
-            } catch (err) {
-                if (err instanceof util.ProtocolError) {
-                    onError(err);
+            ws.onClose((code: number, reason: string) => {
+                // nativescript-websocket on iOS doesn't return error-code 1002 in case of error, but sets the 'reason'
+                // to non-null in case of error.
+                if (code !== 1000 || reason) {
```

Comment:
```suggestion
                if (code !== 1000 || reason !== "") {
```

Reply:
again, I'm not sure the different websocket implementations actually handle it like that.